### PR TITLE
notebooks: store default pattern type per notebook

### PR DESCRIPTION
**Pull request by stefanhengl**
This PR refactors notebooks to support the upcoming Keyword Search GA. The main goal is to make it easier to switch to a new default pattern type without breaking existing notebooks.

**Before**
- pattern type and version were hardcoded in several places

**After**
- Each notebook has a read-only pattern type as determined by the new column `notebooks.pattern_type` (defaults to "standard").

**Notes**
- Notebooks call the Stream API via various helper functions. `patternType` and `version` are both required parameters, which is redundant, because version acts as a default pattern type already. I left a TODO in the code for that. I don't want to change this as part of this PR because the change would get very big and affect too much code outside of Notebooks. 
- We support rendering notebooks with `.snb.md` extension. Unlike notebooks stored in our db, we cannot migrate those files.

**Q&A**
Q: How does this help for Keyword Search GA?
A: Once we default to keyword search, we can change the default of `notebooks.pattern_type` from "standard" to "keyword". Existing notebooks will still work with "standard". New Notebooks will use "keyword". 

Q: How can customers migrate existing notebooks to a new version? 
A: Use the existing "Copy to My Notebooks" function of Notebooks. The copied notebook will have the current default pattern type. 
 
Test plan:
- existing tests pass
- manual testing
  - I created a couple of notebooks with all the different block types and verified via the network tab that all requests to the Stream API have the proper pattern type. I played around with different values in `notebooks.pattern_type` to make sure that the request parameters change.

### CODE REVIEW

File: client/web/src/notebooks/notebook/NotebookComponent.tsx [hunk 1]
```diff
@@ -438,6 +441,7 @@ export const NotebookComponent: React.FunctionComponent<React.PropsWithChildren<
                                 telemetryRecorder={telemetryRecorder}
                                 platformContext={platformContext}
                                 authenticatedUser={authenticatedUser}
+                                queryVersion={queryVersion}
```
Comment (stefanhengl) [inline client/web/src/notebooks/notebook/NotebookComponent.tsx:444]: The "query", "file" and "symbol" blocks each send requests to the Stream API. That's why they al the the query version.

File: client/web/src/notebooks/notebook/index.ts [hunk 2]
```diff
@@ -98,11 +99,13 @@ export class NotebookHeadingMarkdownRenderer extends Renderer {
 export class Notebook {
     private blocks: Map<string, Block>
     private blockOrder: string[]
+    private queryVersion = ''
 
-    constructor(initializerBlocks: BlockInit[], private dependencies: BlockDependencies) {
+    constructor(initializerBlocks: BlockInit[], queryVersion: string, private dependencies: BlockDependencies) {
```
Comment (stefanhengl) [inline client/web/src/notebooks/notebook/index.ts:104]: This is one of the core changes. The Notebook class now has a new property "queryVersion" which we use in the method `runBlockByID`

**Comment by jtibshirani** [comment]
> Once we introduce V4 (not in this PR),

Based on [your Notion doc](https://www.notion.so/sourcegraph/Keyword-Search-GA-a318cec0283a46b384e5b413841986bb), I thought we were not introducing V4 in this first "Phase 1" of work. And that instead, we would persist `patterntype` directly for Insights and Notebooks. Are you now thinking we will introduce V4 and use versions instead?

**Comment by stefanhengl** [comment]
> > Once we introduce V4 (not in this PR),
> 
> Based on [your Notion doc](https://www.notion.so/sourcegraph/Keyword-Search-GA-a318cec0283a46b384e5b413841986bb), I thought we were not introducing V4 in this first "Phase 1" of work. And that instead, we would persist `patterntype` directly for Insights and Notebooks. Are you now thinking we will introduce V4 and use versions instead?
 
For Notebooks, it is not as straightforward to add the pattern type to the query as it is for Code Monitors or Insights. The blocks are multiline, support comments, have autosave, and are collaborative. IMO it is the best experience if the default pattern type is defined per Notebook and not per block. 

Now the question is whether we persist the default pattern type or the version in the db. Functionally, both are the same. I am slightly in favour of using "version" here because it fits to our long term vision to require a "version" whenever we accept or persist a query.

As for the rollout: For "Phase 1", I don't think we should introduce "version" to the API, but we can add support for "V4" to the code without breaking anything.

**Comment by jtibshirani** [comment]
> I am slightly in favour of using "version" here because it fits to our long term vision to require a "version" whenever we accept or persist a query.
> As for the rollout: For "Phase 1", I don't think we should introduce "version" to the API, but we can add support for "V4" to the code without breaking anything.

It feels slightly messy to me to introduce V4 before it's available in the API. We also want "Phase 1" to be self contained and not introduce any tech debt. Plus, Notebooks might be deprecated soon, so it's not a big deal if the schema includes pattern type directly. What do you think?

By the way, this PR looks great other than this question.

**Comment by stefanhengl** [comment]
> > I am slightly in favour of using "version" here because it fits to our long term vision to require a "version" whenever we accept or persist a query.
> > As for the rollout: For "Phase 1", I don't think we should introduce "version" to the API, but we can add support for "V4" to the code without breaking anything.
> 
> It feels slightly messy to me to introduce V4 before it's available in the API. We also want "Phase 1" to be self contained and not introduce any tech debt. Plus, Notebooks might be deprecated soon, so it's not a big deal if the schema includes pattern type directly. What do you think?
> 
> By the way, this PR looks great other than this question.

I swapped version with pattern type. PTAL @jtibshirani

File: migrations/frontend/squashed.sql [hunk 3]
```diff
@@ -76,6 +76,14 @@ CREATE TYPE lsif_uploads_transition_columns AS (
 
 COMMENT ON TYPE lsif_uploads_transition_columns IS 'A type containing the columns that make-up the set of tracked transition columns. Primarily used to create a nulled record due to `OLD` being unset in INSERT queries, and creating a nulled record with a subquery is not allowed.';
 
+CREATE TYPE pattern_type AS ENUM (
+    'keyword',
+    'literal',
+    'regexp',
+    'standard',
+    'structural'
```
Comment (jtibshirani) [inline migrations/frontend/squashed.sql:84]: Would it make sense to only include keyword, regexp, and standard? Can you actually default to structural search in notebooks?

**Review by jtibshirani — APPROVED** [review]
This looks good to me! Thanks for the iterations. I just had one question around the allowed pattern types.

File: migrations/frontend/squashed.sql (hunk 3 above)
Comment (stefanhengl) [reply migrations/frontend/squashed.sql:84]: The default is determined by what we define as default for `notebooks.field pattern_type`, so "no", user are not able to choose the default by themselves. That being said, notebooks do support structural search.

The db type `pattern_type` is not limited to notebooks, that's why I included all non-experimental pattern types.

File: client/shared/src/search/suggestions/index.ts [hunk 4]
```diff
@@ -45,9 +45,18 @@ function firstMatchStreamingSearch(
 }
 
 export function fetchStreamSuggestions(query: string, sourcegraphURL?: string): Observable<SearchMatch[]> {
+    return fetchStreamSuggestionsPatternTypeVersion(query, SearchPatternType.standard, LATEST_VERSION, sourcegraphURL)
+}
+
+export function fetchStreamSuggestionsPatternTypeVersion(
+    query: string,
+    patternType: SearchPatternType,
+    version: string,
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply client/shared/src/search/suggestions/index.ts]: This doesn't really expose "version". The APIs already expose "version" and it is already part of `StreamSearchOptions`. This change merely makes this option available to the helper `fetchStreamSuggestions`. However, it's not critical for this PR, so we can punt on it.

**Pull request merged**